<b><font size="6">Data-Driven Marketing with Google Merch Store Customer Data and Google Play Store App Reviews</font>

    Notebook by Allison Kelly - allisonkelly42@gmail.com

    Blog post - placeholder

    Presentation - placeholder

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span></li><li><span><a href="#Imports" data-toc-modified-id="Imports-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Imports</a></span></li><li><span><a href="#Customer-Lifetime-Value-and-Customer-Attrition-Prediction" data-toc-modified-id="Customer-Lifetime-Value-and-Customer-Attrition-Prediction-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Customer Lifetime Value and Customer Attrition Prediction</a></span><ul class="toc-item"><li><span><a href="#Obtaining-the-data" data-toc-modified-id="Obtaining-the-data-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Obtaining the data</a></span></li><li><span><a href="#Cleaning-the-Data" data-toc-modified-id="Cleaning-the-Data-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Cleaning the Data</a></span></li></ul></li><li><span><a href="#Future-Work" data-toc-modified-id="Future-Work-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Future Work</a></span></li></ul></div>

# Introduction

Leveraging machine learning and AI with the massive amounts of consumer history, web traffic, and product reviews can have a measurable impact on marketing campaigns. By harnessing the power of statistical and analytical tools, patterns untracked by basic dashboards can be unearthed, translating into higher conversion rates over traditional marketing strategies. The following notebook is one such demonstration of this power.

Part one explores customer purchase history on the <a href="https://shop.googlemerchandisestore.com/">Google Merchandise Store.</a> By using machine learning, I will predict the customer lifetime value (CLV) of each customer. This data can then be used to segment ads, email marketing campaigns, and other traditional marketing ventures. I will also predict and measure customer attrition which can be used to develop proactive prevention strategies and increase revenue. 

Part two uses NLP to parse through Google Play store app reviews, conduct sentiment analysis, and classify reviews. The NLP machine learning models can be used to recommend apps and get feedback in realtime to prompt bug fixes and improvements. You'll find this in a separate notebook <b>here.</b> 

Finally, part three will focusing on tracking important metrics and gleaning insights from the Google Analytics demo account for the Google Merchandise Store. Setting benchmarks for KPIs and tracking campaign results are essential to understanding consumer behavior and making informed business decisions. PLACEHOLDER FOR WHERE THIS CAN BE FOUND

# Imports

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns 
import datetime
import numpy as np
import json # for data cleaning
import dill
import pickle

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import metrics

# Customer Lifetime Value and Customer Attrition Prediction

In marketing, according to <a href="https://en.wikipedia.org/wiki/Customer_lifetime_value">Wikipedia</a>, Customer Lifetime Value (CLV) can be defined as "a prediction of the net profit attributed to the entire future relationship with a customer." CLV is an important metric to track as it can help identify your customer acquisition budget, help your marketing team to segment your customer base, increase retention and satisfaction, etc. CLV is represented as a dollar amount and varies customer to customer. 

<img src="brandwise-clv-bellcurve.gif">

## Obtaining the data

This dataset was derived from the Google Merchandise Store demo account as provided by Google BigQuery on Kaggle. You can find the data <a href="https://www.kaggle.com/c/ga-customer-revenue-prediction">here.</a> 

In [2]:
df = pd.read_csv('train.csv') # loading and checking out the dataset
df.head()

/Users/Allie/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


channelGrouping      date  \
0  Organic Search  20160902   
1  Organic Search  20160902   
2  Organic Search  20160902   
3  Organic Search  20160902   
4  Organic Search  20160902   

                                              device        fullVisitorId  \
0  {"browser": "Chrome", "browserVersion": "not a...  1131660440785968503   
1  {"browser": "Firefox", "browserVersion": "not ...   377306020877927890   
2  {"browser": "Chrome", "browserVersion": "not a...  3895546263509774583   
3  {"browser": "UC Browser", "browserVersion": "n...  4763447161404445595   
4  {"browser": "Chrome", "browserVersion": "not a...    27294437909732085   

                                          geoNetwork  \
0  {"continent": "Asia", "subContinent": "Western...   
1  {"continent": "Oceania", "subContinent": "Aust...   
2  {"continent": "Europe", "subContinent": "South...   
3  {"continent": "Asia", "subContinent": "Southea...   
4  {"continent": "Europe", "subContinent": "North...   

                        sessionId  socialEngagementType  \
0  1131660440785968503_1472830385  Not Socially Engaged   
1   377306020877927890_1472880147  Not Socially Engaged   
2  3895546263509774583_1472865386  Not Socially Engaged   
3  4763447161404445595_1472881213  Not Socially Engaged   
4    27294437909732085_1472822600  Not Socially Engaged   

                                              totals  \
0  {"visits": "1", "hits": "1", "pageviews": "1",...   
1  {"visits": "1", "hits": "1", "pageviews": "1",...   
2  {"visits": "1", "hits": "1", "pageviews": "1",...   
3  {"visits": "1", "hits": "1", "pageviews": "1",...   
4  {"visits": "1", "hits": "1", "pageviews": "1",...   

                                       trafficSource     visitId  visitNumber  \
0  {"campaign": "(not set)", "source": "google", ...  1472830385            1   
1  {"campaign": "(not set)", "source": "google", ...  1472880147            1   
2  {"campaign": "(not set)", "source": "google", ...  1472865386            1   
3  {"campaign": "(not set)", "source": "google", ...  1472881213            1   
4  {"campaign": "(not set)", "source": "google", ...  1472822600            2   

   visitStartTime  
0      1472830385  
1      1472880147  
2      1472865386  
3      1472881213  
4      1472822600

In [3]:
print(df.info())

print('\n\nNumber of individual visitors to the Google Merch Store: ', 
      len(df.fullVisitorId.unique()))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903653 entries, 0 to 903652
Data columns (total 12 columns):
channelGrouping         903653 non-null object
date                    903653 non-null int64
device                  903653 non-null object
fullVisitorId           903653 non-null object
geoNetwork              903653 non-null object
sessionId               903653 non-null object
socialEngagementType    903653 non-null object
totals                  903653 non-null object
trafficSource           903653 non-null object
visitId                 903653 non-null int64
visitNumber             903653 non-null int64
visitStartTime          903653 non-null int64
dtypes: int64(4), object(8)
memory usage: 82.7+ MB
None


Number of individual visitors to the Google Merch Store:  742735


The dataframe is consisted of 903,653 datapoints describing visits to the Google Merch Store with 742,753 unique visitors. Most features consist of objects, though the integers are either dates, times or immutable ID numbers that should be categorized accordingly. 

## Cleaning the Data

In [5]:
# try/except or default column names
def datetime_edits(dataframe):
    
    """This function changes the two columns 
    with dates/times into datetime objects.
    To be used with dataframes with similar
    setup to this one."""
    
    try:
        dataframe.date.astype(str)
        dataframe.date = pd.to_datetime(dataframe.date, 
                                        format='%Y%m%d')
    except:
        print("Your dataframe does not contain a 'date' column.")
        
    try:
        dataframe.visitStartTime = pd.to_datetime(dataframe.visitStartTime, unit='s')
    
    except:
        print("Your dataframe does not containt a 'visitStartTime' column.")
    
    # confirming proper edits were made
    return "Date column:", dataframe.date.dtype, "Time column:", dataframe.visitStartTime.dtype


In [6]:
datetime_edits(df)

('Date column:', dtype('<M8[ns]'), 'Time column:', dtype('<M8[ns]'))

In [7]:
# changing immutable IDs to strings
df.visitId = df.visitId.astype(str)

In [8]:

def unpack_col(df, col):
    
    '''Some column values are dictionary-like strings.
    This function unpacks those values, creating new
    columns for each key that can then be analyzed 
    as separate features. Returns dataframe of 
    unpacked column, not including original column.
    '''
    
    df[col] = df[col].dropna().apply(json.loads)
    unpacked_df = df[col].apply(pd.Series)
        
    return unpacked_df
        
device_df = unpack_col(df, 'device')
geo_df = unpack_col(df, 'geoNetwork')
totals_df = unpack_col(df, 'totals')
traffic_df = unpack_col(df, 'trafficSource')

In [9]:
df_expanded = pd.concat([df,device_df, geo_df, totals_df, traffic_df], axis=1)

In [10]:
df_expanded.head()

channelGrouping       date  \
0  Organic Search 2016-09-02   
1  Organic Search 2016-09-02   
2  Organic Search 2016-09-02   
3  Organic Search 2016-09-02   
4  Organic Search 2016-09-02   

                                              device        fullVisitorId  \
0  {'browser': 'Chrome', 'browserVersion': 'not a...  1131660440785968503   
1  {'browser': 'Firefox', 'browserVersion': 'not ...   377306020877927890   
2  {'browser': 'Chrome', 'browserVersion': 'not a...  3895546263509774583   
3  {'browser': 'UC Browser', 'browserVersion': 'n...  4763447161404445595   
4  {'browser': 'Chrome', 'browserVersion': 'not a...    27294437909732085   

                                          geoNetwork  \
0  {'continent': 'Asia', 'subContinent': 'Western...   
1  {'continent': 'Oceania', 'subContinent': 'Aust...   
2  {'continent': 'Europe', 'subContinent': 'South...   
3  {'continent': 'Asia', 'subContinent': 'Southea...   
4  {'continent': 'Europe', 'subContinent': 'North...   

                        sessionId  socialEngagementType  \
0  1131660440785968503_1472830385  Not Socially Engaged   
1   377306020877927890_1472880147  Not Socially Engaged   
2  3895546263509774583_1472865386  Not Socially Engaged   
3  4763447161404445595_1472881213  Not Socially Engaged   
4    27294437909732085_1472822600  Not Socially Engaged   

                                              totals  \
0  {'visits': '1', 'hits': '1', 'pageviews': '1',...   
1  {'visits': '1', 'hits': '1', 'pageviews': '1',...   
2  {'visits': '1', 'hits': '1', 'pageviews': '1',...   
3  {'visits': '1', 'hits': '1', 'pageviews': '1',...   
4  {'visits': '1', 'hits': '1', 'pageviews': '1',...   

                                       trafficSource     visitId     ...       \
0  {'campaign': '(not set)', 'source': 'google', ...  1472830385     ...        
1  {'campaign': '(not set)', 'source': 'google', ...  1472880147     ...        
2  {'campaign': '(not set)', 'source': 'google', ...  1472865386     ...        
3  {'campaign': '(not set)', 'source': 'google', ...  1472881213     ...        
4  {'campaign': '(not set)', 'source': 'google', ...  1472822600     ...        

   transactionRevenue   campaign  source   medium          keyword  \
0                 NaN  (not set)  google  organic   (not provided)   
1                 NaN  (not set)  google  organic   (not provided)   
2                 NaN  (not set)  google  organic   (not provided)   
3                 NaN  (not set)  google  organic  google + online   
4                 NaN  (not set)  google  organic   (not provided)   

                                    adwordsClickInfo isTrueDirect  \
0  {'criteriaParameters': 'not available in demo ...          NaN   
1  {'criteriaParameters': 'not available in demo ...          NaN   
2  {'criteriaParameters': 'not available in demo ...          NaN   
3  {'criteriaParameters': 'not available in demo ...          NaN   
4  {'criteriaParameters': 'not available in demo ...         True   

   referralPath adContent campaignCode  
0           NaN       NaN          NaN  
1           NaN       NaN          NaN  
2           NaN       NaN          NaN  
3           NaN       NaN          NaN  
4           NaN       NaN          NaN  

[5 rows x 54 columns]

In [120]:
df2 = pd.read_csv('expanded_df.csv')
df2.head()

/Users/Allie/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (4,54) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Unnamed: 0 channelGrouping        date  \
0           0  Organic Search  2016-09-02   
1           1  Organic Search  2016-09-02   
2           2  Organic Search  2016-09-02   
3           3  Organic Search  2016-09-02   
4           4  Organic Search  2016-09-02   

                                              device        fullVisitorId  \
0  {'browser': 'Chrome', 'browserVersion': 'not a...  1131660440785968503   
1  {'browser': 'Firefox', 'browserVersion': 'not ...   377306020877927890   
2  {'browser': 'Chrome', 'browserVersion': 'not a...  3895546263509774583   
3  {'browser': 'UC Browser', 'browserVersion': 'n...  4763447161404445595   
4  {'browser': 'Chrome', 'browserVersion': 'not a...    27294437909732085   

                                          geoNetwork  \
0  {'continent': 'Asia', 'subContinent': 'Western...   
1  {'continent': 'Oceania', 'subContinent': 'Aust...   
2  {'continent': 'Europe', 'subContinent': 'South...   
3  {'continent': 'Asia', 'subContinent': 'Southea...   
4  {'continent': 'Europe', 'subContinent': 'North...   

                        sessionId  socialEngagementType  \
0  1131660440785968503_1472830385  Not Socially Engaged   
1   377306020877927890_1472880147  Not Socially Engaged   
2  3895546263509774583_1472865386  Not Socially Engaged   
3  4763447161404445595_1472881213  Not Socially Engaged   
4    27294437909732085_1472822600  Not Socially Engaged   

                                              totals  \
0  {'visits': '1', 'hits': '1', 'pageviews': '1',...   
1  {'visits': '1', 'hits': '1', 'pageviews': '1',...   
2  {'visits': '1', 'hits': '1', 'pageviews': '1',...   
3  {'visits': '1', 'hits': '1', 'pageviews': '1',...   
4  {'visits': '1', 'hits': '1', 'pageviews': '1',...   

                                       trafficSource     ...       \
0  {'campaign': '(not set)', 'source': 'google', ...     ...        
1  {'campaign': '(not set)', 'source': 'google', ...     ...        
2  {'campaign': '(not set)', 'source': 'google', ...     ...        
3  {'campaign': '(not set)', 'source': 'google', ...     ...        
4  {'campaign': '(not set)', 'source': 'google', ...     ...        

   transactionRevenue   campaign  source   medium          keyword  \
0                 NaN  (not set)  google  organic   (not provided)   
1                 NaN  (not set)  google  organic   (not provided)   
2                 NaN  (not set)  google  organic   (not provided)   
3                 NaN  (not set)  google  organic  google + online   
4                 NaN  (not set)  google  organic   (not provided)   

                                    adwordsClickInfo isTrueDirect  \
0  {'criteriaParameters': 'not available in demo ...          NaN   
1  {'criteriaParameters': 'not available in demo ...          NaN   
2  {'criteriaParameters': 'not available in demo ...          NaN   
3  {'criteriaParameters': 'not available in demo ...          NaN   
4  {'criteriaParameters': 'not available in demo ...         True   

  referralPath  adContent campaignCode  
0          NaN        NaN          NaN  
1          NaN        NaN          NaN  
2          NaN        NaN          NaN  
3          NaN        NaN          NaN  
4          NaN        NaN          NaN  

[5 rows x 55 columns]

In [121]:
# Dropping unpacked columns and former index

df2 = df2.drop('Unnamed: 0', axis=1)
df2 = df2.drop('device', axis=1)
df2 = df2.drop('geoNetwork', axis=1)
df2 = df2.drop('totals', axis=1)
df2 = df2.drop('trafficSource', axis=1)

With our unpacked columns, we need to identify and weed out new columns don't contain useful information. For example, it doesn't look like any ad campaigns were drawing visitors, so the columns pertaining to ad campaigns, keywords, etc. are not needed for this particular problem. Below, I've investigated a column that might have needed to be unpacked as well, however the only difference is the gclID parameter which links the Google Analytics account (from which is dataset was derived) to the Google Ads account.

In [122]:
df2.adwordsClickInfo.value_counts().head(3)

{'criteriaParameters': 'not available in demo dataset'}                                                                                                                                                                                             882092
{'page': '1', 'slot': 'Top', 'criteriaParameters': 'not available in demo dataset', 'gclId': 'Cj0KEQjwmIrJBRCRmJ_x7KDo-9oBEiQAuUPKMufMpuG3ZdwYO8GTsjiBFd5MPHStZa9y_9NCrI8X97oaAglc8P8HAQ', 'adNetworkType': 'Google Search', 'isVideoAd': False}        70
{'page': '1', 'slot': 'Top', 'criteriaParameters': 'not available in demo dataset', 'gclId': 'Cj0KEQjw1ee_BRD3hK6x993YzeoBEiQA5RH_BEA562M9tvl_mtnAFvtDnDqOQRp1RvxMMgwjcX1LAfwaAj4o8P8HAQ', 'adNetworkType': 'Google Search', 'isVideoAd': False}        41
Name: adwordsClickInfo, dtype: int64

In [123]:
df2 = df2.drop('adwordsClickInfo', axis=1)

In [124]:
cols = list(df2.columns)

one_variant = []
for col in cols:
    if len(df2[col].unique()) == 1:
        print(df2[col].unique())
        one_variant.append(col)
one_variant

['Not Socially Engaged']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
['not available in demo dataset']
[1]


['socialEngagementType',
 'browserVersion',
 'browserSize',
 'operatingSystemVersion',
 'mobileDeviceBranding',
 'mobileDeviceModel',
 'mobileInputSelector',
 'mobileDeviceInfo',
 'mobileDeviceMarketingName',
 'flashVersion',
 'language',
 'screenColors',
 'screenResolution',
 'cityId',
 'latitude',
 'longitude',
 'networkLocation',
 'visits']

Columns with no variation in value will be dropped. Some columns in a full dataset such as language could be relevant to CLV, but since we don't have that information available through the Google Demo Account, we'll remove those columns. 

In [125]:
df2 = df2.drop(one_variant, axis=1)

Because our unique customer identifier is tied to 'fullVisitorId' and 'visitStartTime' describes the frequency (or lacktherof) of visits to the store, we can safely remove 'sessionId' and 'visitId'.

In [126]:
df2 = df2.drop(['visitId', 'sessionId'], axis=1)

Columns 'region,' 'metro,' and 'city' may have an impact on whether or not a customer will spend in the future and how much due to income levels and cost of living, however most of the information in these categories is not available. We'll drop these columns for this project, though you may consider leaving them if other datasets give a more detailed look into the customers location. We can also drop 'continent' and 'subcontinent' because the country names themselves give us that information inherently. 

In [129]:
df2 = df2.drop(['region', 'metro', 'city', 'continent', 'subContinent' ], axis=1)

In [130]:
# Removing parentheses around "Other" channel group
df2.channelGrouping = df2.channelGrouping.str.strip('()')

In [131]:
# Replacing a variant of NaN with 'Unknown'
df2 = df2.replace('(not set)', 'Unknown')

In [132]:
df2.adContent = df2.adContent.str.replace("[%{}?!/:']", '', regex=True)
df2.adContent = df2.adContent.fillna('None')

In [138]:

df2 = df2.drop(['referralPath', 'campaignCode'], axis=1)

In [141]:
df2.to_csv('cleaned_df.csv')

In [139]:
for col in df2.columns:
    print(df2[col].value_counts().head())

Organic Search    381561
Social            226117
Direct            143026
Referral          104838
Paid Search        25326
Name: channelGrouping, dtype: int64
2016-11-28    4807
2016-11-15    4685
2016-11-14    4466
2016-11-30    4435
2016-10-26    4375
Name: date, dtype: int64
1957458976293878100    262
824839726118485274     245
1856749147915772585    188
3608475193341679870    186
3269834865385146569    152
Name: fullVisitorId, dtype: int64
1    703060
2     92548
3     35843
4     19157
5     11615
Name: visitNumber, dtype: int64
2017-04-25 18:49:35    8
2016-11-05 11:38:24    6
2017-01-17 10:43:22    6
2016-12-10 11:32:05    6
2017-07-24 00:36:42    5
Name: visitStartTime, dtype: int64
Chrome               620364
Safari               182245
Firefox               37069
Internet Explorer     19375
Edge                  10205
Name: browser, dtype: int64
Windows      350072
Macintosh    253938
Android      123892
iOS          107665
Linux         35034
Name: operatingSystem, dtype: 

# Future Work